File: Questions/20_Valid_Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> myStack;
        for (int i = 0; i < s.length(); i++){
            if (s[i] == '{' || s[i] == '(' || s[i] == '[')
                myStack.push(s[i]);
            else if (myStack.empty())
                return false;
            else if (s[i] == '}' && myStack.top() != '{')
                return false;
            else if (s[i] == ')' && myStack.top() != '(')
                return false; 
            else if (s[i] == ']' && myStack.top() != '[')
                return false;
            else 
                myStack.pop();
        }
        return myStack.empty();
    }
};

File: Questions/20_Valid_Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        closeToOpen = {"}": "{", "]": "[", ")": "("}

        # traverse string from right to left
        for c in s:
            if c in closeToOpen:
                # if it is a closing char, we need to have its compliment at the top of the stack
                if not stack or stack[-1] != closeToOpen[c]:
                    return False
                else:
                    stack.pop()
            else:
                stack.append(c)

        return True if not stack else False
